packages/plugin-git/sources/gitUtils.ts

```diff
@@ -317,3 +318,98 @@ export async function clone(url: string, configuration: Configuration) {
+  } while (match === null && nextCwd !== cwd);
+
+  return match;
+}
+
+export async function fetchBase(root: PortablePath, {baseRefs}: {baseRefs: Array<string>}) {
+  if (baseRefs.length === 0)
+    throw new UsageError(`Can't run this command with zero base refs specified.`);
+
+  const ancestorBases = [];
+
+  for (const candidate of baseRefs) {
+    const {code} = await execUtils.execvp(`git`, [`merge-base`, candidate, `HEAD`], {cwd: root});
+    if (code === 0) {
+      ancestorBases.push(candidate);
+    }
+  }
+
+  if (ancestorBases.length === 0)
+    throw new UsageError(`No ancestor could be found between any of HEAD and ${baseRefs.join(`, `)}`);
+
+  const {stdout: mergeBaseStdout} = await execUtils.execvp(`git`, [`merge-base`, `HEAD`, ...ancestorBases], {cwd: root, strict: true});
+  const hash = mergeBaseStdout.trim();
+
+  const {stdout: showStdout} = await execUtils.execvp(`git`, [`show`, `--quiet`, `--pretty=format:%s`, hash], {cwd: root, strict: true});
+  const title = showStdout.trim();
+
+  return {hash, title};
+}
+
+// Note: This returns all changed files from the git diff, which can include
+// files not belonging to a workspace
+export async function fetchChangedFiles(root: PortablePath, {base, project}: {base: string, project: Project}) {
+  const {stdout: localStdout} = await execUtils.execvp(`git`, [`diff`, `--name-only`, `${base}`], {cwd: root, strict: true});
+  const trackedFiles = localStdout.split(/\r\n|\r|\n/).filter(file => file.length > 0).map(file => ppath.resolve(root, npath.toPortablePath(file)));
+
+  const {stdout: untrackedStdout} = await execUtils.execvp(`git`, [`ls-files`, `--others`, `--exclude-standard`], {cwd: root, strict: true});
+  const untrackedFiles = untrackedStdout.split(/\r\n|\r|\n/).filter(file => file.length > 0).map(file => ppath.resolve(root, npath.toPortablePath(file)));
+
+  return [...new Set([...trackedFiles, ...untrackedFiles].sort())];
+}
+
+// Note: yarn artifacts are excluded from workspace change detection
+// as they can be modified by changes to any workspace manifest file.
+export async function fetchChangedWorkspaces({ref, project}: {ref: string | true, project: Project}) {
+  if (project.configuration.projectCwd === null)
+    throw new UsageError(`This command can only be run from within a Yarn project`);
+
+  const ignoredPaths = [
+    ppath.resolve(project.cwd, project.configuration.get(`cacheFolder`)),
+    ppath.resolve(project.cwd, project.configuration.get(`installStatePath`)),
+    ppath.resolve(project.cwd, project.configuration.get(`lockfileFilename`)),
+    ppath.resolve(project.cwd, project.configuration.get(`virtualFolder`)),
+  ];
+  await project.configuration.triggerHook((hooks: Hooks) => {
+    return hooks.populateYarnPaths;
+  }, project, (path: PortablePath | null) => {
+    if (path != null) {
+      ignoredPaths.push(path);
+    }
+  });
+
+  const root = await fetchRoot(project.configuration.projectCwd);
+  const base = root !== null
```

Comment:
Should this instead throw if the root could not be found? That would simplify a lot of the branching and it's possibly the right thing to do: why would you use `--since` outside of a Git repo?

Reply:
Would make sense indeed; perhaps in a later diff

Reply:
Good idea. Fixed in 711507b73bf4475858bdcd05d4d4aa4bb6bd5335.